modules/mono/glue/GodotSharp/GodotSharp/Core/Variant.conv.cs

```diff
@@ -0,0 +1,407 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using Godot.NativeInterop;
+
+namespace Godot;
+
+public partial struct Variant
+{
+    private static Exception UnsupportedType<T>() => throw new InvalidOperationException(
+        $"The type is not supported for conversion to/from Variant: '{typeof(T).FullName}'");
+
+    internal static class GenericConversion<T>
+    {
+        public static unsafe godot_variant ToVariant(in T from) =>
+            ToVariantCb != null ? ToVariantCb(from) : throw UnsupportedType<T>();
+
+        public static unsafe T FromVariant(in godot_variant variant) =>
+            FromVariantCb != null ? FromVariantCb(variant) : throw UnsupportedType<T>();
+
+        // ReSharper disable once StaticMemberInGenericType
+        internal static unsafe delegate*<in T, godot_variant> ToVariantCb;
+
+        // ReSharper disable once StaticMemberInGenericType
+        internal static unsafe delegate*<in godot_variant, T> FromVariantCb;
+
+        [SuppressMessage("ReSharper", "RedundantNameQualifier")]
+        static GenericConversion()
+        {
+            RuntimeHelpers.RunClassConstructor(typeof(T).TypeHandle);
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    [SuppressMessage("ReSharper", "RedundantNameQualifier")]
+    public static godot_variant From<[MustBeVariant] T>(in T from)
```

Comment:
> The `in` keyword will certainly be worse for reference types and small value types but if it's inlined it doesn't matter, right? So the `in` keyword does nothing here, I think.

Can't really change it without affecting the other types, but it's inlined so I think either way it's fine (although I didn't test without `in` so I don't know if it would be the same). I prefer `in` in this case, as it's clearer about parameter being immutable.

> Also, are you sure we want a static `From` method in `Variant` that returns a `godot_variant`? Just mentioning it because a static method named `From` is essentially a "constructor" so it feels a little weird that it returns `godot_variant` instead of `Variant`.
> 
> And I guess `Variant` is more of a public-facing type and I expect users to be somewhat familiar with it but not `godot_variant` so it _might_ be confusing. In fact, I guess `godot_variant` would be internal if we didn't need to use it in source generators.

I forgot this was returning `godot_variant`. I'm moving it to `VariantUtils` then.

Reply:
Moved it to `VariantUtils`. Left helper methods in `Variant` as well, in case somebody needs them.